llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -440,14 +442,187 @@ def construct_blocks(self, config: MPTConfig) -> nn.ModuleList:
             nn.ModuleList: The list of Transformer blocks.
         """
         block_args = self.extract_block_args(config.to_dict())
+        self.kv_cache_layers = None
+
+        if config.block_overrides is not None:
+            block_args_list = self._get_override_block_args_list(
+                config,
+                block_args,
+            )
+        else:
+            block_args_list = [block_args for _ in range(config.n_layers)]
 
         return nn.ModuleList([
             self.block_class(
                 device=config.init_device,
-                **block_args,
-            ) for _ in range(config.n_layers)
+                **block_args_i,
+            ) for block_args_i in block_args_list
         ])
 
+    def _get_override_block_args_list(
+        self,
+        config: MPTConfig,
+        block_args: Dict[str, Any],
+    ) -> List[Dict[str, Any]]:
+        if config.block_overrides is None:
+            raise ValueError(
+                'config.block_overrides should not be None when calling _get_override_block_args_list.',
+            )
+        repeat = config.block_overrides[
+            'repeat'] if 'repeat' in config.block_overrides else 1
+        model_modules_order_expanded = MPTModel._get_modules_order_expanded(
+            config.block_overrides['order'],
+        ) * repeat
+        if len(model_modules_order_expanded) != config.n_layers:
+            raise ValueError(
+                f'The specified block overrides do not match the number of layers: {len(model_modules_order_expanded)} vs {config.n_layers}.',
+            )
+
+        new_block_args_list = []
+        layer_description_list = []
+
+        reuse_kv_layer_idx_dict = {}
+        for b_idx in range(config.n_layers):
+            module_name = model_modules_order_expanded[b_idx]
+            override_config = {}
+            if module_name != 'default':
+                override_config = copy.deepcopy(
+                    config.block_overrides['overrides'][module_name],
+                )
+                if 'reuse_kv_layer_idx' in override_config.get(
+                    'attn_config',
+                    {},
+                ):
+                    MPTModel._validate_reuse_kv_layer_config(
+                        block_overrides=config.block_overrides,
+                        model_modules_order_expanded=
+                        model_modules_order_expanded,
+                        b_idx=b_idx,
+                        override_config=override_config,
+                        reuse_kv_layer_idx_dict=reuse_kv_layer_idx_dict,
+                    )
+                    if self.kv_cache_layers is None:
```

Comment:
should this just be initialized to a set to start with?

Reply:
Done.